crates/ruff_linter/src/message/rdjson.rs

```diff
@@ -0,0 +1,138 @@
+use std::io::Write;
+
+use serde::ser::SerializeSeq;
+use serde::{Serialize, Serializer};
+use serde_json::{json, Value};
+
+use ruff_diagnostics::Edit;
+use ruff_source_file::SourceCode;
+use ruff_text_size::Ranged;
+
+use crate::message::{Emitter, EmitterContext, Message, SourceLocation};
+use crate::registry::AsRule;
+
+#[derive(Default)]
+pub struct RdjsonEmitter;
+
+impl Emitter for RdjsonEmitter {
+    fn emit(
+        &mut self,
+        writer: &mut dyn Write,
+        messages: &[Message],
+        _context: &EmitterContext,
+    ) -> anyhow::Result<()> {
+        serde_json::to_writer_pretty(
+            writer,
+            &json!({
+                "source": {
+                    "name": "ruff",
+                    "url": "https://docs.astral.sh/ruff",
+                },
+                "severity": "warning",
+                "diagnostics": &ExpandedMessages{ messages }
+            }),
+        )?;
+
+        Ok(())
+    }
+}
+
+struct ExpandedMessages<'a> {
+    messages: &'a [Message],
+}
+
+impl Serialize for ExpandedMessages<'_> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut s = serializer.serialize_seq(Some(self.messages.len()))?;
+
+        for message in self.messages {
+            let value = message_to_rdjson_value(message);
+            s.serialize_element(&value)?;
+        }
+
+        s.end()
+    }
+}
+
+fn message_to_rdjson_value(message: &Message) -> Value {
+    let source_code = message.file.to_source_code();
+
+    let start_location = source_code.source_location(message.start());
+    let end_location = source_code.source_location(message.end());
+
+    if let Some(fix) = message.fix.as_ref() {
+        json!({
+            "message": message.kind.body,
+            "location": {
+                "path": message.filename(),
+                "range": rdjson_range(&start_location, &end_location),
+            },
+            "code": {
+                "value": message.kind.rule().noqa_code().to_string(),
+                "url": message.kind.rule().url(),
+            },
+            "suggestions": rdjson_suggestions(fix.edits(), &source_code),
+        })
+    } else {
+        json!({
+            "message": message.kind.body,
+            "location": {
+                "path": message.filename(),
+                "range": rdjson_range(&start_location, &end_location),
+            },
+            "code": {
+                "value": message.kind.rule().noqa_code().to_string(),
+                "url": message.kind.rule().url(),
+            },
+        })
```

Comment:
I decided to just duplicate the repeated fields rather than have to unwrap and mutate. (Sort of an arbitrary decision, this is just my preferred style.)